### ISSUE TITLE
EPIC: Confirm WCAG Level II Accessibility Standards Compliance

### ISSUE THREAD

**Issue by ajpeddakotla** [hypothesis/product-backlog#1085]
This is an update of a previous epic: https://github.com/hypothesis/product-backlog/issues/753

### Problem 
An increasing number of schools and institutions are requiring that any tool they adopt must meet basic standards of accessibility. In order for these institutions to adopt Hypothesis is any capacity, we therefore need to meet WCAG Level II standards. 

**Out of scope for this work are the following areas:** 

- Account creation process
- Installation of the Bookmarklet and Chrome extension
- Activation of Bookmarklet and Chrome extension
- Activity pages

### Current Status
@klemay and I have performed a review of the current state of accessibility across our app. Since we've started this process, a number of improvements have been made, some of which have rendered issues in the [Accessibility Roadmap](https://github.com/hypothesis/product-backlog/projects/7) as out of date. This epic describes the remaining work needed to meet WCAG Level II accessibility standards. 

### Issues reported in previous audit

A [previous audit](https://docs.google.com/document/d/1LjiZ5yFSbRZ7WLgKQJ3K7-Xga311efY1zyypHAFMPhQ/edit#heading=h.tyjcwt) reported a number of issues which need to be addressed, including in the areas of:

- [1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG21/#info-and-relationships): 
--- iFrame Titles: we've completed this. 
--- Show/Hide Content Toggles: Unclear
--- Radio Buttons: We think this is good to go
--- Pop-Up elements: We think this is good as well
--- Tabs: We need to double check this
--- Aria-Labels for tags: In Progress
--- Input Errors: Unclear

- [4.1.2 Name, Role, Value](https://www.w3.org/TR/WCAG21/#name-role-value)
Needs Developer evaluation. This evaluation can be done at same time as 1.3.1

- [1.4.1 Use of Color](https://www.w3.org/TR/WCAG21/#use-of-color) (https://github.com/hypothesis/product-backlog/issues/785)
We haven't done this one. What's needed is another way to identify annotated content beyond just highlights. For example, underline or use icons, or textures to identify highlighted and annotated content. 

- [1.4.3 Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum)
This is in progress. Developers need to do an audit to determine where contrast areas have not been addressed. 

- [1.4.4 Resize text](https://www.w3.org/TR/WCAG21/#resize-text) (https://github.com/hypothesis/product-backlog/issues/788)
It works for HTML but this fails for PDFs. You currently cannot resize Hypothesis on PDFs. If you try to change the zoom of the screen, only the PDF changes, and not H. 

- [2.1.1 Keyboard](https://www.w3.org/TR/WCAG21/#keyboard) (https://github.com/hypothesis/product-backlog/issues/447)
This is in progress. Developers need to do an audit to determine the areas that have not been addressed. 

- [2.4.7 Focus Visible](https://www.w3.org/TR/WCAG21/#focus-visible) (https://github.com/hypothesis/product-backlog/issues/447)
Developers can audit this along with the evaluation of 2.1.1. to determine the areas that have not been addressed. 

- [2.2.1 Timing Adjustable ](https://www.w3.org/TR/WCAG21/#timing-adjustable)
We need to evaluate if we have sessions that expire in less than 20 hours. If we do log out users after 20 hours then we need to create a mechanism for a user to override the log out such that we warn them that we'll be logging them out and they can stay logged in if they'd like. 

- [4.1.1 Parsing](https://www.w3.org/TR/WCAG21/#parsing)
Needs Developer evaluation

### Next Steps
- [x]  Developers to evaluate each of the criteria and determine the work needed to meet criteria.
- [x]  Create issues to track remaining work

### Sub-tasks

(Updated 2020-03-17 by @lyzadanger)

- [x] Enable jsx-a11y ESLint plugin for Hypothesis client (https://github.com/hypothesis/client/pull/1728)
- [x] Enable jsx-a11y ESLint plugin for LMS frontend (https://github.com/hypothesis/lms/pull/1378)
- [x] Fix accessibility issues found by jsx-a11y ESLint plugin in Hypothesis client (https://github.com/hypothesis/client/issues/1727)
- [x] Fix accessibility issues found by jsx-a11y ESLint plugin in LMS frontend (https://github.com/hypothesis/lms/issues/1377)
- [x] Implement a11y tests using axe-core in Hypothesis client (https://github.com/hypothesis/client/pull/1729)
- [x] Fix accessibility issues found by axe-core in Hypothesis client (https://github.com/hypothesis/client/issues/1734)
- [x] Implement a11y tests using axe-core in LMS frontend (https://github.com/hypothesis/lms/pull/1395)
- [x] Fix accessibility issues found by axe-core in LMS frontend (https://github.com/hypothesis/lms/issues/1396)
- [x] Install axe browser extension for Chrome and use it to audit HTML and PDF Hypothesis assignments in the LMS app. File issues for anything not covered by the automatic checks above
- [x] Fix issues arising from the above tests
- [x] Audit and address contrast issues in the client per [1.4.3 Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) (https://github.com/hypothesis/client/issues/1809)
- [x] Review WCAG issues found by [previous external a11y audit](https://docs.google.com/document/d/1LjiZ5yFSbRZ7WLgKQJ3K7-Xga311efY1zyypHAFMPhQ/edit#heading=h.tyjcwt) and file issues for violations which still exist
- [x] Schedule another round of a11y external auditing?

- [x] Figure out how creating annotations with the keyboard is going to work and implement it (https://github.com/hypothesis/product-backlog/issues/447)
- [x] Figure out how to make highlights "visible" to users who can't see them and implement it (https://github.com/hypothesis/product-backlog/issues/785)
- [x] Re-implement toast messages in the client and make them accessible (https://github.com/hypothesis/client/issues/1925)
- [x] Make it possible to filter the annotation list in the sidebar to those associated with the text currently selected in the host document (i.e. "clicking" on an existing highlight) (https://github.com/hypothesis/client/issues/1928)
- [x] Make sure that menus implement standard keyboard and focus patterns (https://github.com/hypothesis/client/issues/1926)
- [x] Improving grouping and semantic markup for threads and annotation cards to make them easier to perceive and navigate (https://github.com/hypothesis/client/issues/1927)

Tasks that go beyond pure compliance, but are important:

- [x] Do manual testing of core workflows on HTML and PDF assignments in the LMS app and file issues for anything not caught by the above which would prevent screen reader users from completing them (User Story along the lines of "as a user who relies on a screenreader, I can complete my homework assignment that relies on Hypothesis")
- [x] Fix any issues arising from review of previous audit

**Comment by klemay** [comment]
Jeremy has added the specific contract language we've employed re: accessibility to Slack: https://hypothes-is.slack.com/archives/C8TPC8XMK/p1578338205002900
